prog/mutation.go

```diff
@@ -383,23 +384,18 @@ func (t *BufferType) mutate(r *randGen, s *state, arg Arg, ctx ArgCtx) (calls []
 	return
 }
 
-var imageMu sync.Mutex
-
-func (r *randGen) mutateImage(image []byte) (data []byte, retry bool) {
-	if len(image) == 0 {
-		return image, true
-	}
-	// Don't decompress more than one image at a time
-	// since it can consume lots of memory.
-	// Reconsider when/if we move mutation to the host process.
-	imageMu.Lock()
-	defer imageMu.Unlock()
-	data, err := Decompress(image)
-	if err != nil {
-		panic(fmt.Sprintf("could not decompress data: %v", err))
-	}
+func (r *randGen) mutateImage(compressed []byte) (data []byte, retry bool) {
+	data, dtor := image.MustDecompress(compressed)
+	defer dtor()
 	if len(data) == 0 {
-		return image, true // Do not mutate empty data.
+		return compressed, true // Do not mutate empty data.
+	}
+	if r.oneOf(10) {
+		if len(data) > 1 && r.bin() {
+			data = data[:r.randRange(1, uint64(len(data)))]
+		} else if len(data) < image.MaxSize {
+			data = data[:r.randRange(uint64(len(data)), image.MaxSize)]
```

Comment:
It relies on `data` being actually `image.MaxSize` size, right? If yes, is it the case for [compression_nonoptimized.go](https://github.com/google/syzkaller/pull/3561/files#diff-1d1aecbab5adaf7ed798e519540e433e7ad3307ae11ba578260c66929ec882f1)?

Reply:
You are right. Need to fix.
I started with mmap for 32-bit systems as well (but with a global mutex), but it did not work for windows. Windows has mmap analog, but I don't want to spend time on that.

Reply:
I've dropped the commit for now.
I've stashed it here:
https://github.com/dvyukov/syzkaller/commits/dvyukov-image-size-mutation
but let's deal with it later.
PTAL